x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/esql/action/EsqlQueryResponse.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.esql.action;
+
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.core.Releasable;
+
+import java.util.Iterator;
+import java.util.List;
+
+public abstract class EsqlQueryResponse extends ActionResponse implements Releasable {
+
+    public abstract List<? extends ColumnInfo> columns();
```

Comment:
One thing to add is having some basic typeness to the response. The column type is a string but it for numeric and strings, we could also have a class that we can pass on as a type token:
```
ColumnInfo(name:String, type: String, Class<T> classType) --> default Object.class

public abstract <T> Iterable<T> column(int index, Class<T> type);
```

so folks could do something like response.column(1, columns.get(1).classType());
It's a bit out of scope though.

Reply:
classType is something that I thought about too. And we have elementType in the engine. Lemme see if we can fit it here without too much fuss.

Reply:
Let's leave this to a follow-on iteration.